jstz_api/src/text_encoder.rs

```diff
@@ -1,71 +1,304 @@
 
 #[derive(Trace, Finalize)]
-struct TextEncoder;
+pub struct TextEncoder;
 
-impl TextEncoder {
-    fn atob(data: &JsString) -> JsResult<JsString> {
-        fn on_err(err: impl Error) -> JsError {
-            JsNativeError::eval().with_message(err.to_string()).into()
-        }
-        let str = data.to_std_string_escaped();
-        let encoded = DEFAULT_ENGINE.decode(&str).map_err(on_err)?;
-        let encoded_str = core::str::from_utf8(encoded.as_slice()).map_err(on_err)?;
+#[derive(Trace, Finalize, Default, TryFromJs)]
+pub struct TextEncoderEncodeIntoResult {
+    read: u128,
+    written: u128,
+}
 
-        Ok(encoded_str.into())
+impl TextEncoderEncodeIntoResult {
+    fn try_from_js<'a>(value: &'a JsValue) -> JsResult<GcRefMut<'a, Object, Self>> {
+        value
+            .as_object()
+            .and_then(|obj| obj.downcast_mut::<Self>())
+            .ok_or_else(|| {
+                JsNativeError::typ()
+                    .with_message("Failed to convert js value into rust type `TextEncoderEncodeIntoResult`")
+                    .into()
+            })
+    }
+    fn read(context: &mut Context<'_>) -> Accessor {
+        accessor!(
+            context,
+            TextEncoderEncodeIntoResult,
+            "read",
+            get:((x, _context) => Ok(JsBigInt::new(x.read).into()))
+        )
     }
-    fn btoa(data: &JsString) -> JsResult<JsString> {
-        let str = data.to_std_string_escaped();
-        let encoded = DEFAULT_ENGINE.encode(&str);
-        Ok(encoded.into())
+    fn written(context: &mut Context<'_>) -> Accessor {
+        accessor!(
+            context,
+            TextEncoderEncodeIntoResult,
+            "written",
+            get:((x, _context) => Ok(JsBigInt::new(x.written).into()))
+        )
     }
 }
 
-pub struct TextEncoderApi;
-impl TextEncoderApi {
-    const NAME: &str = "TextEncoder";
-    fn atob(_: &JsValue, args: &[JsValue], _: &mut Context) -> JsResult<JsValue> {
-        let data = args
-            .get_or_undefined(0)
-            .as_string()
-            .ok_or_else(|| JsNativeError::typ().with_message("expected string"))?;
-        let result = TextEncoder::atob(data)?;
-        Ok(result.into())
+impl NativeClass for TextEncoderEncodeIntoResult {
+    type Instance = TextEncoderEncodeIntoResult;
+
+    const NAME: &'static str = "TextEncoderEncodeIntoResult";
+
+    fn constructor(
+        _this: &JsNativeObject<TextEncoderEncodeIntoResult>,
+        _args: &[JsValue],
+        _: &mut Context<'_>,
+    ) -> JsResult<TextEncoderEncodeIntoResult> {
+        Ok(TextEncoderEncodeIntoResult::default())
+    }
+
+    fn init(class: &mut ClassBuilder<'_, '_>) -> JsResult<()> {
+        let read = Self::read(class.context());
+        let written = Self::written(class.context());
+        class
+            .accessor(js_string!("read"), read, Attribute::all())
+            .accessor(js_string!("written"), written, Attribute::all());
+
+        Ok(())
+    }
+}
```

Comment:
Let's keep this for now and add a TODO comment

Wrt to dictionaries, given their copying semantics when being passed / returned in the WHATWG spec, in future they should be Rust structs that implement `TryFromJs` and `TryIntoJs`.

Reply:
done